### PR TITLE
L3 loc fast

### PR DESCRIPTION
**Pull request by gregthelaw**

### CODE REVIEW

File: l3.S [hunk 1]
```diff
@@ -9,20 +9,26 @@ l3__log_fast:
...
-    mov %rdi, (%rcx)  // msg
-    add $8, %rcx
-    movq $0, (%rcx)  // arg1
-    add $8, %rcx
-    movq $0, (%rcx)  // arg1
+    add $32, %r8
+    shl $5, %r9
+    add %r9, %r8
+    mov %eax, (%r8)  // tid
+#ifdef L3_LOC_ENABLED
```
Comment (gapisback) [inline l3.S:22]: @gregthelaw - There is something funny going on with this logic of stashing away the `__LOC__ ID` value, which should only happen if `L3_LOC_ENABLED=1`. Otherwise, I expect only 0 to be stashed away.

I will explain this as a comment in the top-level UI. One has to xcheck the outputs from CI jobs.

Comment (gapisback) [inline l3.S:22]: Are we allowed to have blank lines and // comments in assembly code? If so, I think this block might merit some commentary.

.S code is hard to follow as-it-is, so if you could leave a blank linke before this line and add some // comments, it will be helpful for future readers.

File: l3.S [hunk 2]
```diff
@@ -9,20 +9,26 @@ l3__log_fast:
...
+    add $4, %r8
+    mov %esi, (%r8)  // loc
+    add $4, %r8
+#else
+    add $8, %r8
+#endif
```
Comment (gapisback) [inline l3.S:28]: Nothing major, but ... 

I like to stick to proper form of tagging the #else on L26 and this one with `// L3_LOC_ENABLED`

It may be obvious now, but helps with overall readability.

**Review by gapisback — CHANGES_REQUESTED** [review]
I think some more investigation is needed to figure out exactly what's being stashed away by the `l3_fast_log()` interface.

If you see this output [Build-and-Run-C-Samples-with-LOC](https://github.com/undoio/l3/actions/runs/8545488811/job/23413905764?pr=17), see this chunk:

```
python3 l3_dump.py /tmp/l3.c-test.dat ./build/release/bin/use-cases/single-file-C-program
tid=3388 single-file-C-program/test-main.c:76  '300-Mil Simple l3-log msgs' arg1=299998464 arg2=0
tid=3388 single-file-C-program/test-main.c:76  '300-Mil Simple l3-log msgs' arg1=299998465 arg2=0
tid=3388 single-file-C-program/test-main.c:76  '300-Mil Simple l3-log msgs' arg1=299998466 arg2=0
```

The output is consistent with this block of code from the test-main.c:

```
 75     for (n = 0; n < (nMil * L3_MILLION); n++) {
 76         l3_log_simple("300-Mil Simple l3-log msgs", n, 0);
 77     }
```
So, that's good; this case works fine.

----

But if you see this output, [Build-and-Run-Cpp-Samples-with-LOC](https://github.com/undoio/l3/actions/runs/8545488811/job/23413905764?pr=17), i.e., .cpp code with `L3_LOC_ENABLED=1`, we see this output:

```
python3 l3_dump.py /tmp/l3.cpp-test.dat ./build/release/bin/use-cases/single-file-Cpp-program
tid=19850 single-file-Cpp-program/test-main.cpp:51  '300-Mil Fast l3-log msgs' arg1=47140 arg2=24
tid=19850 single-file-Cpp-program/test-main.cpp:51  '300-Mil Fast l3-log msgs' arg1=47140 arg2=24
tid=19850 single-file-Cpp-program/test-main.cpp:51  '300-Mil Fast l3-log msgs' arg1=47140 arg2=24
```

But in `use-cases/single-file-Cpp-program/test-main.cpp`, the code looks like so, for simple logging:

```
 50     for (n = 0; n < (nMil * L3_MILLION); n++) {
 51         l3_log_simple("300-Mil Simple l3-log msgs", 0, 0);
 52     }
```

And like so, below, for fast-logging:

```
 67     for (n = 0; n < (300 * L3_MILLION); n++) {
 68         l3_log_fast("300-Mil Fast l3-log msgs");
 69     }
```

In the fast-log trace output, we are getting the right message, but the wrong line number. 

It is still somehow saving-off the line # from the call to `l3_log_simple()` from L51, even when `l3_log_fast()` is executed from L68.

I think the assembly is, perhaps, not over-writing the new `loc_t loc` value supplied by this expansion for fast-logging; from l3.h:

```
 82 #ifdef L3_LOC_ENABLED
 83 #define l3_log_fast(msg) l3__log_fast((msg), __LOC__)
 84 #else   // L3_LOC_ENABLED
 85 #define l3_log_fast(msg) l3__log_fast((msg), L3_ARG_UNUSED)
 86 #endif  //
```

----

The same issue is seen for the other `.cc test-main.cc` as well:

Outputs from `Build-and-Run-CC-Samples-with-LOC`:

The line number reported in this output:

```
python3 l3_dump.py /tmp/l3.cc-test.dat ./build/release/bin/use-cases/single-file-CC-program
tid=36311 single-file-CC-program/test-main.cc:51  '300-Mil Fast l3-log msgs' arg1=56948 arg2=24
tid=36311 single-file-CC-program/test-main.cc:51  '300-Mil Fast l3-log msgs' arg1=56948 arg2=24
tid=36311 single-file-CC-program/test-main.cc:51  '300-Mil Fast l3-log msgs' arg1=56948 arg2=24
```

... does not match this line # in the code `~/Projects/undoio-l3/use-cases/single-file-CC-program/test-main.cc`:

```
 67     for (n = 0; n < (300 * L3_MILLION); n++) {
 68         l3_log_fast("300-Mil Fast l3-log msgs");
 69     }
```

I expect to see L68 coming out of decoded LOC-ID, not L 51 (which is for `l3_simple_log()` logging).

----

Overall, I think your assembly code is doing something reasonable, but possibly there are some sneaking-bugs.

I don't follow the .S changes, so you will need to decipher this when you have some cycles.

**Comment by gapisback** [comment]
@gregthelaw , I noticed one more thing with the debug output. For e.g., in this we see:

```
tid=19850 single-file-Cpp-program/test-main.cpp:51  '300-Mil Fast l3-log msgs' arg1=47140 arg2=24
```

But in this `test-main.cpp` file we are doing:

```
 68         l3_log_fast("300-Mil Fast l3-log msgs");
```

No args are being supplied, so if anything I'd have expected to see value of 0 for arg1, arg2 in above output.

Is your `.S` change trying to support supplying args? If so, the wrapper macro might also want to change, below:

```
 82 #ifdef L3_LOC_ENABLED
 83 #define l3_log_fast(msg) l3__log_fast((msg), __LOC__)
 84 #else   // L3_LOC_ENABLED
 85 #define l3_log_fast(msg) l3__log_fast((msg), L3_ARG_UNUSED)
 86 #endif  //
```

**Comment by gapisback** [comment]
@gregthelaw : Just in the hope of helping out here, I rebased the base branch of this branch, i.e., `gapisback/fast_log-LOC-support` off of `main`. And rebased this branch off of that `gapisback/fast_log-LOC-support`.

I re-pushed this dev-branch to github, so your dev-branch should now be in-sync with `/main.`

File: l3.S (hunk 1 above)
Comment (gregthelaw) [reply l3.S:20]: I think I figured it out - the L3_LOC_ENABLED macro wasn't finding its way to the .S file. I've pushed a commit to this branch which updates the Makefile.

Comment (gregthelaw) [reply l3.S:20]: Good thinking -- done.

File: l3.S (hunk 2 above)
Comment (gregthelaw) [reply l3.S]: done

File: l3.S [hunk 3]
```diff
@@ -2,27 +2,31 @@
 .extern l3_log
 
 l3__log_fast:
-    mov %fs:0xfffffffffffffffc,%eax
+    mov %fs:0xfffffffffffffffc,%eax // did we already stuff the tid in TLS?
```
Comment (gapisback) [inline l3.S:5]: Most comments are declarative, but this one is asking a question.

I think it may be simpler, and easier to follow, if this comment is also phrased like it's on L8 for the other `mov` command.

Truly, I don't even understand what this assembly does -- so will leave it up to you to rephrase this as a statement rather than a question.

**Review by gapisback — APPROVED** [review]
@gregthelaw -- THANK YOU, for adding copious comments!

I still don't follow the assembly, but these comments will certainly help the next person (most likely yourself) who comes along on this `.S` file to figure stuff out.

I went back and x-checked the CI-job outputs. My suspicions reported earlier of the incorrect line#s being printed are seemingly resolved now. 

From a spot-check, the `.S` changes seem to be doing the right thing.

Once we have this in `/main`, I'll write-up unit-pytests to cross-check with a smaller workload, to ensure that the right `loc-ID` is being stashed away in all execution modes.

**Comment by gapisback** [comment]
@gregthelaw -- As this the 1st time we are collaborating on a common dev-branch, I want to know how you want to move this forward.

I would like to push the changes from your branch into my dev-branch `gapisback/fast_log-LOC-support`, where I will integrate this work with other in-flight work. I'll probably squash all commits into one coherent commit, add any [more] unit-tests, if needed, and will surface that PR #16 for a review back to you.

At that time, you will rubber-stamp this `.S` work and just review the remaining changes I've added for integrating fast-logging w/LOC-ID support.

Hope this works for you.

**Comment by gapisback** [comment]
@gregthelaw - Sorry to drag this on. Previously, I reviewed for correctness of the line # stashed away as LOC-ID by this fast-logging support.

With your newest `.S` work, are we supposed to save-off the actual argument values, or 0?

In this file, `use-cases/single-file-CC-program/test-main.cc`, I see this chunk:

```
 67     for (n = 0; n < (300 * L3_MILLION); n++) {
 68         l3_log_fast("300-Mil Fast l3-log msgs");
 69     }
```

Here, it appears that no args are being supplied, so I expect to see 0s in the output.

But in the CI-output for this run, say, `Build-and-Run-CC-Samples-with-LOC`, I see this output:

```
python3 l3_dump.py /tmp/l3.cc-test.dat ./build/release/bin/use-cases/single-file-CC-program
tid=36281 single-file-CC-program/test-main.cc:68  '300-Mil Fast l3-log msgs' arg1=61002 arg2=24
tid=36281 single-file-CC-program/test-main.cc:68  '300-Mil Fast l3-log msgs' arg1=61002 arg2=24
```

Why am I still seeing non-zero args; i.e. `arg1=61002 arg2=24`?

Seems like some left-over cruft is creeping through?

I think this will be confusing. Either we fully support saving-off args which, I think, will be much more useful in the longer-run (even though at a small perf cost). Or, we save-off only 0s.

You don't need to resolve this right now, as you are caught up with travel etc.

If you agree with my findings, I can log an open issue and we can come back to it later, along with a new pytest case.

**Comment by gregthelaw** [comment]
Oops, the assembly code was stashing the 'arguments' but the C wrappers weren't passing them in. I've pushed a commit that fixes this.

**Comment by gapisback** [comment]
Thanks, @gregthelaw for this correction.

I re-ran the use-case samples and the output with fast-logging seems to be coming correctly on all fronts.

One small last request. It's do with the the interface of this new fast-logging, which looks like:

```
 #ifdef L3_LOC_ENABLED
-#define l3_log_fast(msg) l3__log_fast((msg), __LOC__)
+#define l3_log_fast(msg, arg1, arg2) l3__log_fast((msg), __LOC__, (arg1), (arg2))
 #else   // L3_LOC_ENABLED
-#define l3_log_fast(msg) l3__log_fast((msg), L3_ARG_UNUSED)
+#define l3_log_fast(msg, arg1, arg2) l3__log_fast((msg), L3_ARG_UNUSED, (arg1), (arg2))
 #endif  //
```

I think you just extended this based on what I attempted in earlier commits.

It will be better to line-up the interface exactly with what's out there for simple-logging:

```
     71 #ifdef L3_LOC_ENABLED
     72 void l3__log_simple(loc_t loc, const char *msg,
     73                     const uint64_t arg1, const uint64_t arg2);
     74 #else
     75 void l3__log_simple(const uint32_t loc, const char *msg,
     76                     const uint64_t arg1, const uint64_t arg2);
     77 #endif  // L3_LOC_ENABLED
```

----

Can you change the interface:

from: ` void l3__log_fast(const char *msg, loc_t loc, const uint64_t arg1, const uint64_t arg2);`
   to :  ` void l3__log_fast(loc_t loc, const char *msg, const uint64_t arg1, const uint64_t arg2);`

It will involve a bit of roto-tilling to fix up the caller-macros and the test code.

I'd do this myself, post-check-in, but as this involves tweaking assembly code (which I'm not comfortable with), I'm requesting you to do this clean-up while this change-set is still in your dev-branch.

Let me know when this is done, I'll plan to push this to my dev-branch, squash all commits and will air out the whole change-set for a final review by you.
